torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,609 @@
+      gtd_prev = gtd_new;
+      obj_func_res = obj_func(x, t, d);
+      f_new = std::get<0>(obj_func_res);
+      g_new = std::get<1>(obj_func_res);
+      ls_func_evals += 1;
+      gtd_new = g_new.dot(d);
+      ls_iter += 1;
+    }
+    // reached max number of iterations?
+    if (ls_iter == max_ls) {
+      bracket = {0, t};
+      bracket_f = {f, f_new};
+      bracket_g = {g, g_new};
+    }
+
+    // zoom phase: we now have a point satisfying the criteria, or
+    // a bracket around it. We refine the bracket until we find the
+    // exact point satisfying the criteria
+    bool insuf_progress = false;
+    // find high and low points in bracket
+    int64_t low_pos, high_pos;
+    if (bracket_f[0] <= bracket_f[1]) {
+      low_pos = 0;
+      high_pos = 1;
+    } else {
+      low_pos = 1;
+      high_pos = 0;
+    }
+    while(!done && (ls_iter < max_ls)) {
+      // compute new trial value
+      t = _cubic_interpolate(bracket[0], bracket_f[0], val(bracket_gtd[0]),
+                             bracket[1], bracket_f[1], val(bracket_gtd[1]));
+
+      // test that we are making sufficient progress:
+      // in case `t` is so close to boundary, we mark that we are making
+      // insufficient progress, and if
+      //   + we have made insufficient progress in the last step, or
+      //   + `t` is at one of the boundary,
+      // we will move `t` to a position which is `0.1 * len(bracket)`
+      // away from the nearest boundary point.
+      double bracket_max = std::max(bracket[0], bracket[1]);
+      auto bracket_min = std::min(bracket[0], bracket[1]);
+      auto eps = 0.1 * (bracket_max - bracket_min); // // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+      if (std::min(bracket_max - t, t - bracket_min) < eps) {
+        // interpolation close to boundary
+        if (insuf_progress || (t >= bracket_max) || (t <= bracket_min)) {
+          // evaluate at 0.1 away from boundary
+          t = (abs(t - bracket_max) < abs(t - bracket_min)) ? bracket_max - eps : bracket_min + eps;
+          insuf_progress = false;
+        } else {
+          insuf_progress = true;
+        }
+      } else {
+        insuf_progress = false;
+      }
+
+      // Evaluate new point
+      obj_func_res = obj_func(x, t, d);
+      f_new = std::get<0>(obj_func_res);
+      g_new = std::get<1>(obj_func_res);
+      ls_func_evals += 1;
+      gtd_new = g_new.dot(d);
+      ls_iter += 1;
+
+      if ((f_new > (f + c1 * t * val(gtd))) || (f_new >= bracket_f[low_pos])) {
+        // Armijo condition not satisfied or not lower than lowest point
+        // # Armijo condition not satisfied or not lower than lowest point
+        bracket[high_pos] = t;
+        bracket_f[high_pos] = f_new;
+        bracket_g[high_pos] = g_new.clone(at::MemoryFormat::Contiguous);
+        bracket_gtd[high_pos] = gtd_new;
+        low_pos = (bracket_f[0] <= bracket_f[1]) ? 0 : 1;
+        high_pos = (bracket_f[0] <= bracket_f[1]) ? 1 : 0;
```

Comment:
might be easier to write:
```suggestion
        std::tie(low_pos, high_pos) = bracket_f[0] <= bracket_f[1] ? std::make_tuple(0, 1) : std::make_tuple(1, 0);
```